src/server/frontend_wayland/text_input_v3.cpp

```diff
@@ -292,6 +323,41 @@ void mf::TextInputV3::send_text_change(ms::TextInputChange const& change)
         // We are no longer enabled or we don't have a valid serial
         return;
     }
+    if (change.keysym)
+    {
+        auto nano = std::chrono::milliseconds{change.keysym->time};
+        keyboard_device->if_started_then([&](input::InputSink* sink, input::EventBuilder* builder)
+            {
+                auto state = change.keysym->state;
+                auto scan_code = change.keysym->sym;
+                switch (scan_code)
+                {
+                    case XKB_KEY_BackSpace:
+                        scan_code = KEY_BACKSPACE;
+                        break;
+                    case XKB_KEY_Return:
+                        scan_code = KEY_ENTER;
+                        break;
+                    case XKB_KEY_Left:
+                        scan_code = KEY_LEFT;
+                        break;
+                    case XKB_KEY_Right:
+                        scan_code = KEY_RIGHT;
+                        break;
+                    case XKB_KEY_Up:
+                        scan_code = KEY_UP;
+                        break;
+                    case XKB_KEY_Down:
+                        scan_code = KEY_DOWN;
+                        break;
+                    default:
+                        scan_code = KEY_UNKNOWN;
+                        break;
+                }
```

Comment:
Maybe extract this as a utility function?

```suggestion
                auto const scan_code = sym_to_scan_code(change.keysym->sym);
```